docs/architecture/adr-042-state-sync.md

```diff
@@ -0,0 +1,209 @@
+chunks and snappy compressed. Hashes of snappy compressed chunks are stored in a
+manifest file which co-ordinates the state-sync. Obtaining a correct manifest
+file seems to require an honest majority of peers. This means you may not find
+out the state is incorrect until you download the whole thing and compare it
+with a verified block header.
+
+A similar solution was implemented by Binance in
+[#3594](https://github.com/tendermint/tendermint/pull/3594)
+based on their initial implementation in
+[PR #3243](https://github.com/tendermint/tendermint/pull/3243)
+and [some learnings](https://docs.google.com/document/d/1npGTAa1qxe8EQZ1wG0a0Sip9t5oX2vYZNUDwr_LVRR4/edit).
+Note this still requires the honest majority peer assumption.
+One major advantage of the warp-sync approach is that chunks can be compressed before sending.
+
+Ideally State Sync in Tendermint would not require an honest majority of peers, and instead
+would rely on the standard light client security, meaning state chunks would be
+verifiable against the merkle root of the state. Such a design for tendermint
+was originally tracked in [#828](https://github.com/tendermint/tendermint/issues/828).
+
+### Analysis of Lazy vs Eager
+
+Lazy vs Eager have more in common than they differ. They all require
+reactors on the tendermint side, a set of ABCI messages and a method for
+serializing/deserializing snapshots facilitated by a SnapshotFormat.
+
+The biggest difference between Lazy and Eager proposals is in the
+read/write patterns necessitated by when serving a snapshot chunk.
+Specifically, Lazy State Sync random reads to the underlying data
+structure while Eager can optimize for sequential reads.
+
+This distinctin between approaches was demonstrated by Binance's [ackratos](https://github.com/ackratos) in their
+implementation of [Lazy State sync](https://github.com/tendermint/tendermint/pull/3243), The [analysis](https://docs.google.com/document/d/1npGTAa1qxe8EQZ1wG0a0Sip9t5oX2vYZNUDwr_LVRR4/) of the performance, and
+follow up implementation of [Warp Sync](http://github.com/tendermint/tendermint/pull/3594).
+
+One proposed difference between Lazy and Eager state sync is the safety
+model. It had been suggested that Lazy State Sync could use light client
+validation while Warp Sync was unsuitable for public networks as it
+relies on the majority of peers to produce a manifest. This manifest
+would require complete syncing before validation of the data against the
+merkle root were possible. This could create DDOS attack vector in which
+a peer publishes an invalid manifest and making peers download invalid
+data. For this reason, WarpSync is faster but less safety-efficient; It
+as it takes longer to realize that data is invalid.
+
+However, upon analysis of the approaches with the orthogonal axis of
+eager/lazy and chunk/snapshot validation suggest that a version of
+WarpSync with per chunk light client validation might get the optimal
```

Comment:
Maybe something like:

```
While warp sync in general does not permit continuous light-client validation of chunks
(since chunks are typically agnostic of underlying tree structure), 
it may be possible to provide this by taking care to chunk the state in a way that preserves 
enough structure of the tree to allow chunks to be fully validated on receipt. 
Thus an Eager approach that can still harness light-client validation may get the optimal performance and safety.
```

Reply:
Are we really sure that other implementations of warp sync don't let you validate chunks until the very end?

Reply:
I don't understand what is meant with WarpSync approach here. Maybe it's better explaining the properties of what we see as an optimal design without referring to existing approaches with different designs as this just creates confusion. For example, optimal design would rely on light client verification for verifying root hash of the snapshot state and provide possibility for chunk based validation with respect to the root hash.

Reply:
Chunks will contain some parts of underlying tree structure, right (keys & values)? As long as the client & server agree on the chunking algorithm I don't see why random parts of each chunk (key-value pairs) can't be validated occasionally, but maybe I'm missing something.